tests/integration/test_integration.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+import logging
+from pathlib import Path
+
+import pytest
+import yaml
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+
+TLS_REQUIRER_CHARM_NAME = "tls-certificates-requirer"
+
+
+@pytest.fixture(scope="module")
+@pytest.mark.abort_on_fail
+async def build_and_deploy(ops_test: OpsTest):
+    """Build the charm-under-test and deploy it."""
+    charm = await ops_test.build_charm(".")
+    resources = {"lego-image": METADATA["resources"]["lego-image"]["upstream-source"]}
+    assert ops_test.model
+    await ops_test.model.deploy(
+        charm,
+        resources=resources,
+        application_name=APP_NAME,
+        series="jammy",
+        config={
+            "email": "example@email.com",
+            "aws_access_key_id": "dummy key",
+            "aws_secret_access_key": "dummy access key",
+            "aws_region": "dummy region",
+            "aws_hosted_zone_id": "dummy zone id",
+        },
+    )
+    await ops_test.model.deploy(
+        TLS_REQUIRER_CHARM_NAME,
+        application_name=TLS_REQUIRER_CHARM_NAME,
+        channel="edge",
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_given_charm_is_built_when_deployed_then_status_is_active(
+    ops_test: OpsTest,
+    build_and_deploy,
+):
+    assert ops_test.model
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        timeout=1000,
+    )
+
+
+async def test_given_tls_requirer_is_deployed_and_related_then_status_is_active(
+    ops_test: OpsTest,
+    build_and_deploy,
+):
+    assert ops_test.model
+    await ops_test.model.add_relation(
+        relation1=f"{APP_NAME}:certificates", relation2=f"{TLS_REQUIRER_CHARM_NAME}"
+    )
+    await ops_test.model.wait_for_idle(
+        apps=[TLS_REQUIRER_CHARM_NAME],
+        status="active",
+        timeout=1000,
+    )
```

Comment:
I think this test works just accidentally. The provider cannot request a certificate and errors out, but the status of the requirer is still active. For all the lego operators, I think we will have to make the test a bit more complex if we want them to have any value.

Reply:
Yeah right now we more or less only validate that both side have a `certificates` interface. If we want to validate that an actual cert has been emitted we'll need to provide AWS credentials which I wanted to avoid here.